src/org/sosy_lab/java_smt/example/NQueens.java

```diff
@@ -0,0 +1,208 @@
+    abstract Boolean getValue(BooleanFormula[][] symbols, Model model, int row, int col);
+
+    /**
+     * Solves the N-Queens problem for the given board size and returns a possible solution.
+     * Returns <code>Null</code> if no solution exists.
+     */
+    @Nullable
+    public Boolean[][] solve(int n) throws InterruptedException, SolverException {
+        BooleanFormula[][] symbols = getSymbols();
+        List<BooleanFormula> rules = getRules(symbols, context);
+        // solve N-Queens
+        try (ProverEnvironment prover = context.newProverEnvironment(ProverOptions.GENERATE_MODELS)) {
+            prover.push(bmgr.and(rules));
+            boolean isUnsolvable = prover.isUnsat();
+            if (isUnsolvable) {
+                return null;
+            }
+
+            // get model and convert it
+            Boolean[][] solution = new Boolean[n][n] ;
+            try (Model model = prover.getModel()) {
+                for (int row = 0; row < n; row++) {
+                    for (int col = 0; col < n; col++) {
+                        solution[row][col] = getValue(symbols, model, row,col);
+                    }
+                }
+                return solution;
+            }
+        }
+    }
+}
+class NQueen extends NQueensSolver {
+    private final int n;
+
+    public NQueen(SolverContext context, int n) {
+        super(context);
+        this.n = n;
+    }
+
+    /**
+     * prepare symbols: one symbol for each of the N*N cells.
+     */
+    @Override
+    BooleanFormula[][] getSymbols() {
+        final BooleanFormula[][] symbols = new BooleanFormula[n][n];
+        for (int row = 0; row < n; row++) {
+            for (int col = 0; col < n; col++) {
+                symbols[row][col] = bmgr.makeVariable("q_" + row + "_" + col);
+            }
+        }
+        return symbols;
+    }
+
+    /*
+     * getRules is the method used to add constraints that ensure that no two queens are in the same
+     * row, column, or diagonal.
+     */
+    @Override
+    List<BooleanFormula> getRules(BooleanFormula[][] symbols, SolverContext context) {
+        List<BooleanFormula> rules = new ArrayList<>();
+
+        // Add constraints to ensure that only one queen is placed in each row
+        for (int row = 0; row < n; row++) {
+            List<BooleanFormula> ors = new ArrayList<>();
+            for (int col = 0; col < n; col++) {
+                ors.add(symbols[row][col]);
+            }
+            rules.add(bmgr.or(ors));
+        }
+        // Add constraints to ensure that only one queen is placed in each column
+        for (int col = 0; col < n; col++) {
+            List<BooleanFormula> ors = new ArrayList<>();
+            for (int row = 0; row < n; row++) {
+                ors.add(symbols[row][col]);
+            }
+            rules.add(bmgr.or(ors));
+        }
+
+        // Add constraints to ensure that only one queen is placed in each diagonal
+        for (int i = 0; i < n; i++) {
```

Comment:
This constraint requires that there is at least one queen in every diagonal, or even more.
This is wrong and needs to be improved, such there can be at most one queen per diagonal (even no queen at all is allowed).